gk/main.c

```diff
@@ -408,7 +433,8 @@ gk_process_granted(struct flow_entry *fe, struct ipacket *packet)
 	if (ret < 0)
 		return ret;
 
-	/* TODO Fill up the Ethernet header of the packet. */
+	eth_hdr = rte_pktmbuf_mtod(packet->pkt, struct ether_hdr *);
+	rte_memcpy(eth_hdr, &eth_cache->eth_hdr, sizeof(*eth_hdr));
```

Comment:
I think `encapsulate()` only prepends enough space to the packet buffer to add the outer L3 header, but does not leave enough room for the Ethernet header. I think the call to `rte_pktmbuf_prepend()` in `encapsulate()` needs to be changed.